lightning/src/ln/channel.rs

```diff
@@ -1717,8 +1765,28 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 				removed_outbound_total_msat += htlc.amount_msat;
 			}
 		}
-		if htlc_inbound_value_msat + msg.amount_msat + self.value_to_self_msat > (self.channel_value_satoshis - Channel::<ChanSigner>::get_our_channel_reserve_satoshis(self.channel_value_satoshis)) * 1000 + removed_outbound_total_msat {
-			return Err(ChannelError::Close("Remote HTLC add would put them over their reserve value"));
+
+		if self.channel_outbound {
+			// Check that they won't violate our channel reserve by adding this HTLC.
+
+			// One of the +1's is for the HTLC that is currently being added to the commitment tx
+			// and the other is a fee spike buffer we're keeping for the remote (this deviates
+			// from the spec but should help protect us from stuck channels).
+			// TODO: is the reason this added HTLC's amount can't count toward's the receiver's
+			// fee spike buffer because that doesn't work with existing HTLC output spend scripts?
+			let local_fee_cost_msat = self.commit_tx_fee_msat(self.htlc_count_next_local_commit_tx() + 1 + 1);
+			if self.value_to_self_msat < self.their_channel_reserve_satoshis * 1000 + local_fee_cost_msat {
+				return Err(ChannelError::Ignore("Cannot receive value that would put us over their reserve value"));
+			}
+		}
+
+		// The +1 is for the HTLC that is currently being added to the commitment tx.
+		let remote_fee_cost_msat = if self.channel_outbound { 0 } else {
+			self.commit_tx_fee_msat(self.htlc_count_next_remote_commit_tx() + 1)
+		};
+		if htlc_inbound_value_msat + msg.amount_msat + self.value_to_self_msat + remote_fee_cost_msat
+			> (self.channel_value_satoshis - Channel::<ChanSigner>::get_our_channel_reserve_satoshis(self.channel_value_satoshis)) * 1000 + removed_outbound_total_msat {
```

Comment:
>get_our_channel_reserve_satoshis is terribly confusing
Totally agree, reopened #181 and self-assigned

Reply:
> It took me quite some time to reason about the entire expression. It may be more intuitive if formulated using variables that are in terms of the remote party (i.e., the one responsible for paying fees), each of which encompasses some of the arithmetic. Consider the following, possibly renaming variables:
> 
> ```rust
> let pending_value_to_self_msat =
>     self.value_to_self_msat + htlc_inbound_value_msat - removed_outbound_total_msat;
> let pending_remote_value_msat =
>     self.channel_value_satoshis * 1000 - pending_value_to_self_msat;
> let channel_reserve_msat =
>     Channel::<ChanSigner>::get_our_channel_reserve_satoshis(self.channel_value_satoshis) * 1000;
> 
> if pending_remote_value_msat - msg.amount_msat - remote_fee_cost_msat < channel_reserve_msat {
>     // ...
> }
> ```
> 
> I believe that is an equivalent expression (modulo allowing to be _at_ the channel reserve). Could possibly fold ` msg.amount_msat` into one of the variables to make the condition more compact.

@jkczyz This does seem much more readable. Just a heads up, I'm going to put the subtracted values as additions on the other side to prevent overflow (and add a corresponding comment). See this comment https://github.com/rust-bitcoin/rust-lightning/pull/577#discussion_r407825688 for explanation